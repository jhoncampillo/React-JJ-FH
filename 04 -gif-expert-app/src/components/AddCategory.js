import React, { useState } from "react";
import PropTypes from "prop-types";

//importo el estado
//puedo trae la props setCatecories con solo 'props'
export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState(""); //strig vacio

  //creo la funcion para capturar el evento e.target.value
  //se lo asigono al setInputValue
  //con el evento onChage cambio el contenido del input
  const handleInputChangue = (e) => {
    console.log(e.target.value);
    setInputValue(e.target.value);
  };

  //creo en evento enter
  const handleSubmit = (e) => {
    //quito el refresh del navegador
    e.preventDefault();
    console.log("Submit Echo");

    //avaluoo si son espcios
    if (inputValue.trim().length > 2) {
      //console.log(porsp.setCategories); asi seria con props
      //comono tengo acceso al inputValue c
      //adiciojo el inpul a los cata\s que ya hay
      setCategories((cats) => [...cats, inputValue]);
      setInputValue("");
    }
  };
  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChangue}
      ></input>
    </form>
  );
};
//Defino el props types requiered
AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
