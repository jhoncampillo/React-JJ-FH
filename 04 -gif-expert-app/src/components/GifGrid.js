import { useState } from "react";
import React from "react";
//recive el argumento category que viene de input del AddCategory
export const GifGrid = ({ category }) => {
  //creo el estado
  const [count, setCount] = useState(0);

  const getGifs = async () => {
    const url =
      "https://api.giphy.com/v1/gifs/search?q=Goku&limit=10&api_key=rRKAHmsjH1qgSc0SUfBItcBaut4L7DUQ";
    const respuesta = await fetch(url);
    //console.log(respuesta);
    //desetructuring solo traigo la data data.data
    const { data } = await respuesta.json();

    //la recorro con map
    const gifs = data.map((img) => {
      return {
        id: img.id,
        title: img.title,
        url: img.images.downsized_medium.url,
      };
    });
    console.log(gifs);
  };
  getGifs();
  return (
    <div>
      <h3>{category} </h3>
      <h3>{count}</h3>
      <button onClick={() => setCount(count + 1)}></button>
    </div>
  );
};
