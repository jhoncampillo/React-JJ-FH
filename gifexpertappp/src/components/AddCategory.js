import React from "react";
import { useState } from "react";
import PropTypes from "prop-types";
//Traigo el parametro entre corchetes por el desetructuring
const AddCategory = ({ setCategories }) => {
  //creo el useState Para saber llo que la persona escribe en el input

  const [inputValue, setInputValue] = useState(""); //Seimpre es string Vacio

  //creo la funcion del onChangue
  const handleInputChangue = (e) => {
    setInputValue(e.target.value);

    //Log para pruebas
    console.log("handleInputChangue Llamado");
  };

  //CUANDO PULSA ENTER
  const handleSubmit = (e) => {
    //quita el refresh del navegador
    e.preventDefault();
    console.log("Submit OK- Handle Prueba 3 ", { inputValue });

    //valido espacios en blaco con trim
    if (inputValue.trim().length > 2) {
      //ADICIONO EL VALOR DEL INPUT AL LA <li></li>
      //Pondo de primera la ultima categoria solicitada
      setCategories((cats) => [inputValue, ...cats]);
      //reseteo en setinputValue
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      {/* paramtro para Prueba */}
      <p>{inputValue}</p>

      {/* <h2>Addcategory</h2> */}
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChangue}
        placeholder="Busca tu Personaje"
      />
    </form>
  );
};
//Defino el parametro setCategories como required
AddCategory.propTypes = {
  //defino una funcion como required
  setCategories: PropTypes.func.isRequired,
};

export default AddCategory;
