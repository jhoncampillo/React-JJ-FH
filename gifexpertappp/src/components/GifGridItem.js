import React from "react";
//recibo las propiedades del componente GigGrid = id, title, url
//export const GifGridItem = (props) => {
//importo propytypes para props obligatorias - imtp+tab
import PropTypes from "prop-types";

export const GifGridItem = ({ id, title, url }) => {
  //console.log(props); remplazo las propuiedades por id,title,url
  // console.log({ id, title, url });
  return (
    <div className="card animate__animated animate__backInRight">
      {/* {img.title} 
        descargo las imagenes en el DOM
      */}
      <img src={url} alt={title} />
      <p className="title">{title}</p>
    </div>
  );
};
//Defino proptypes Obligatorios
GifGridItem.prototype = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
