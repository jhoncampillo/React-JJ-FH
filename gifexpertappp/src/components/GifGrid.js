import React from "react";
//import { useEffect, useState } from "react";
import { GifGridItem } from "./GifGridItem";
//import { getGifs } from "../helpers/getGifs";
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from "prop-types";

//Recibo el parametro category de GifexpertApp con Destructuring
const GifGrid = ({ category }) => {
  //creo el use state - pero sin el useEffect se ejecuta la funcion en
  //cada evento - se vuelve infinita.- se dispara cada vez que hay un cambio
  //const [images, setImages] = useState([]);

  //USO EL CUSTOM HOOKS
  //const state = useFetchGifs();
  //desestruturing-  data: images = renombra data con images
  const { loading, data: images } = useFetchGifs(category);

  ///console.log(loading);
  //console.log(data);

  //necesito recibir la categoria y hacer una peticion http

  //creo un afuncion

  //getGifs();
  //Capitalizo
  const Category = category.charAt(0).toUpperCase() + category.slice(1);
  console.log(Category);
  return (
    <>
      <h2>{Category}</h2>
      {/* //uso el Custom hooks */}
      {/* //&& si es verdadero muetras de lo contratrio no muestra nada. */}
      {loading && (
        <p className="animate__animated animate__bounceIn">Loading...</p>
      )}
      <div className="card-grid">
        {/* <h3>{count}</h3> */}

        {/* <button onClick={() => setCount(count + 1)}></button> */}

        {/* //terer en cuenta que el return implicito  va con () 
          Puedo desestructurar este map
          */}
        {/* {images.map((img) => (
          <li key={img.id}>{img.title}</li>
        ))} // Le mando el nuevo componente*/}
        {images.map((img) => (
          //   <li key={id}>{title}</li>
          <GifGridItem
            key={img.id}
            // img={img} replazo esto y mando al componente todas las propiedades de img
            //como independientes id,title,url
            {...img}
          />
        ))}
      </div>
    </>
  );
};

//Defino que el el parametro category es requerido
GifGrid.protoTypes = {
  category: PropTypes.string.isRequired,
};

export default GifGrid;
