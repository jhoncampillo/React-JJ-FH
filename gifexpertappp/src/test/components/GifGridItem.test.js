import React from "react";
import { shallow } from "enzyme";
import { GifGridItem } from "../../components/GifGridItem";

describe("Pruebas del Componente <GifgridItems>", () => {
  //let wrapper = shallow(<GifGridItem />);
  //mandar los proptypes
  const title = "Un titulo";
  const url = "https://loclahosts.com/";
  //mando las props al wraper del componente-
  //Subro el scope del warpper para que sea global
  const wrapper = shallow(<GifGridItem title={title} url={url} />);
  //Primer test
  test("Prueba de Mostar Correctamente GifGridItems", () => {
    //wrapper = shallow(<GifGridItem />);

    expect(wrapper).toMatchSnapshot();
  });

  //segundo test

  test("Debe de terner un parrfo con el titulo ", () => {
    //Busco el parrafo
    const p = wrapper.find("p");
    expect(p.text().trim()).toBe(title);
  });

  //Nueba prueba
  test("Debe tener una imagen igual al  url y alt de los props ", () => {
    //busco la imagen
    const img = wrapper.find("img");
    //Observo el resultado dela busqueda con find
    console.log(img.html());
    console.log(img.props()); //prop(src) evalua una sola propeidad
    console.log(img.props().src);
    //espera que la url de ma imagen se igual a la declarada

    expect(img.prop("src")).toBe(url);
    expect(img.prop("alt")).toBe(title);
  });
  //TEST PARA BUSCAR CLASES
  test("debe tener la classname animate __backInRight", () => {
    //busco el div
    const div = wrapper.find("div");

    console.log(div.prop("classname"));
    //asigno a la const classname el valor del div.prop
    const className = div.prop("className");
    //evalue que sea true osea que exista
    expect(className.includes("animate__backInRight")).toBe(true);
  });
});
