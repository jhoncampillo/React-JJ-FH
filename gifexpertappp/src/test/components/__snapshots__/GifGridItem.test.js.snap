// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del Componente <GifgridItems> Prueba de Mostar Correctamente GifGridItems 1`] = `
<div
  className="card animate__animated animate__backInRight"
>
  <img
    alt="Un titulo"
    src="https://loclahosts.com/"
  />
  <p
    className="title"
  >
    Un titulo
  </p>
</div>
`;
